### ISSUE TITLE
RNN-T greedy search cannot emit >1 symbol per time-step

### ISSUE THREAD

**Issue by xkr1** [espnet/espnet#2196]
Regarding the implementation of the greedy decoder, I believe the following if statement should be replaced with a while loop:

https://github.com/espnet/espnet/blob/0fe5068a356f5e48beac25117fa1d6c2ad799452/espnet/nets/pytorch_backend/transducer/rnn_decoders.py#L201

The current implementation looks like it allows only one symbol to be emitted at each time step.

At each time step, if a non-blank symbol is the most probable symbol, the state of the prediction network should be advanced and the output distribution re-evaluated. If a non-blank symbol is the most probable from this re-evaluated output distribution we should emit that too, and continue this process until we encounter a case where the blank symbol has the highest probability (or some emission threshold is hit - determined by a new hyperparameter to prevent you getting stuck endlessly emitting symbols at a particular timestep in badly conditioned cases). Only then should we move to the next time step.

Was this a design decision for efficiency reasons, or is this a bug? I'm thinking of a case with low-frame rate models with short-duration output symbols (i.e. graphemes) that may rely on emitting multiple symbols per time step (i.e. a grapheme model will typically emit "t" and "h" together in a single timestep when recognising "the", "there", etc.)

**Comment by b-flo** [comment]
> Was this a design decision for efficiency reasons, or is this a bug?

This is intended! I didn't observe any issues with that restriction though. Did you?
It's also a restriction applied in several beam search variant.

**Comment by xkr1** [comment]
I've certainly seen cases where multiple symbols are emitted on a single time step in other implementations. It may be possible to recover on the next time step in this implementation, but it will depend very much on the type of output units and frame-rate of the model.

In the case that zero or one symbols are emitted on each time step the while-loop implementation should be equivalent to the current implementation. It's only when more symbols are emitted that it gets slower, but arguably you want those symbols to be emitted so it's an acceptable increase in computational complexity. 

The model is trained using a loss function that permits transitions that you currently don't allow in decoding - if the restrictions in decoding do not degrade performance, I wonder if it's possible to modify the loss function to make that more efficient to compute as well?

You refer to other beam search variants having the same restriction - would you be so kind as to post some references to those, please? I'm still learning about these models and keen to learn more.

**Comment by b-flo** [comment]
> I've certainly seen cases where multiple symbols are emitted on a single time step in other implementations. It may be possible to recover on the next time step in this implementation, but it will depend very much on the type of output units and frame-rate of the model.

From what we experimented the performance/decoding time trade-off was worth the limitation and caused no issues. However,  I'm always open to modify/upgrade any part related to transducer! I just need some feedback or someone to run some experiments as most of my ressources are allocated on other things for now :)

> The model is trained using a loss function that permits transitions that you currently don't allow in decoding - if the restrictions in decoding do not degrade performance, I wonder if it's possible to modify the loss function to make that more efficient to compute as well?

There are several interesting techniques to train rnn-transducer (or transformer-transducer) more efficiently, such as [monotonic rnn-t](https://ieeexplore.ieee.org/document/9003822), [modified encoder and prediction output combination](https://www.microsoft.com/en-us/research/uploads/prod/2019/10/RNNT.pdf), etc.
Some things are on the todo list but well, I dunno when it will be available.

> You refer to other beam search variants having the same restriction - would you be so kind as to post some references to those, please? I'm still learning about these models and keen to learn more.

Sure, sorry! For example: [one step constrained beam search](https://arxiv.org/abs/2002.03577), [breadth-first beam search](https://ieeexplore.ieee.org/document/9003822) and [time synchronous beam search](https://ieeexplore.ieee.org/document/9053040). The last two ones don't strictly enfore to only one expansion step, the number of allowed expansions at each time step can be controlled by the user.

**Comment by xkr1** [comment]
Thanks for the links!

In a different implementation of RNN-T greedy search (my own Tensorflow-based decoder) I noticed that there were times when multiple symbols are emitted at the same time step. When I get a chance I'll modify my implementation to your method and compare the final WER on Librispeech. The findings should be equally applicable to your implementation. Will get back to you soon.
